validate/josev2/validator_test.go

```diff
@@ -0,0 +1,160 @@
+			keyFunc: func(context.Context) (interface{}, error) {
+				return []byte("secret"), nil
+			},
+			expectedContext: &UserContext{
+				RegisteredClaims: jwt.Claims{
+					Issuer:   issuer,
+					Subject:  subject,
+					Audience: jwt.Audience{audience},
+				},
+			},
+		},
+		{
+			name:  "it successfully validates a token with custom claims",
+			token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczovL2dvLWp3dC1taWRkbGV3YXJlLmV1LmF1dGgwLmNvbS8iLCJzdWIiOiIxMjM0NTY3ODkwIiwiYXVkIjpbImh0dHBzOi8vZ28tand0LW1pZGRsZXdhcmUtYXBpLyJdLCJzY29wZSI6InJlYWQ6bWVzc2FnZXMifQ.oqtUZQ-Q8un4CPduUBdGVq5gXpQVIFT_QSQjkOXFT5I",
+			keyFunc: func(context.Context) (interface{}, error) {
+				return []byte("secret"), nil
+			},
+			customClaims: &testClaims{},
+			expectedContext: &UserContext{
+				RegisteredClaims: jwt.Claims{
+					Issuer:   issuer,
+					Subject:  subject,
+					Audience: jwt.Audience{audience},
+				},
+				CustomClaims: &testClaims{
+					Scope: "read:messages",
+				},
+			},
+		},
+		{
+			name:  "it throws an error when token has a different signing algorithm than the validator",
+			token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczovL2dvLWp3dC1taWRkbGV3YXJlLmV1LmF1dGgwLmNvbS8iLCJzdWIiOiIxMjM0NTY3ODkwIiwiYXVkIjpbImh0dHBzOi8vZ28tand0LW1pZGRsZXdhcmUtYXBpLyJdfQ.-R2K2tZHDrgsEh9JNWcyk4aljtR6gZK0s2anNGlfwz0",
+			keyFunc: func(context.Context) (interface{}, error) {
+				return []byte("secret"), nil
+			},
+			algorithm:     jose.RS256,
+			expectedError: errors.New(`expected "RS256" signing algorithm but token specified "HS256"`),
+		},
+		{
+			name:  "it throws an error when it cannot parse the token",
+			token: "",
+			keyFunc: func(context.Context) (interface{}, error) {
+				return []byte("secret"), nil
+			},
+			expectedError: errors.New("could not parse the token: square/go-jose: compact JWS format must have three parts"),
+		},
+		{
+			name:  "it throws an error when it fails to fetch the keys from the key func",
+			token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczovL2dvLWp3dC1taWRkbGV3YXJlLmV1LmF1dGgwLmNvbS8iLCJzdWIiOiIxMjM0NTY3ODkwIiwiYXVkIjpbImh0dHBzOi8vZ28tand0LW1pZGRsZXdhcmUtYXBpLyJdfQ.-R2K2tZHDrgsEh9JNWcyk4aljtR6gZK0s2anNGlfwz0",
+			keyFunc: func(context.Context) (interface{}, error) {
+				return nil, errors.New("key func error message")
+			},
+			expectedError: errors.New("error getting the keys from the key func: key func error message"),
+		},
+		{
+			name:  "it throws an error when it fails to deserialize the claims",
+			token: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczovL2dvLWp3dC1taWRkbGV3YXJlLmV1LmF1dGgwLmNvbS8iLCJzdWIiOiIxMjM0NTY3ODkwIiwiYXVkIjpbImh0dHBzOi8vZ28tand0LW1pZGRsZXdhcmUtYXBpLyJdfQ.vR2K2tZHDrgsEh9zNWcyk4aljtR6gZK0s2anNGlfwz0",
```

Comment:
Just a question, why in this particular case the claims fail to be deserialized, when they seem correct (syntax & expected type)?
The error seems to be related to the signature.

Reply:
I updated the test name to better reflect why we fail to deserialize the claims.

Reply:
https://github.com/auth0/go-jwt-middleware/pull/119/commits/14e3dc69a6f743b40504778ca4788446aac038dd

Reply:
Also changed the signature of the validator.New to not use any jose pkg structs.